src/algebra/module/bimodule.lean

```diff
@@ -0,0 +1,129 @@
+to `S`, this approach is more general.
+
+## TODO
+
+Develop the theory of two-sided ideals, which have type `submodule (R ⊗[ℕ] Rᵐᵒᵖ) R`.
+
+-/
+
+open_locale tensor_product
+
+local attribute [instance] tensor_product.algebra.module
+
+namespace subbimodule
+
+section algebra
+
+variables {R A B M : Type*}
+variables [comm_semiring R] [add_comm_monoid M] [module R M]
+variables [semiring A] [semiring B] [module A M] [module B M]
+variables [algebra R A] [algebra R B]
+variables [is_scalar_tower R A M] [is_scalar_tower R B M]
+variables [smul_comm_class A B M]
+
+/-- A constructor for a subbimodule which demands closure under the two sets of scalars
+individually, rather than jointly via their tensor product.
+
+Note that `R` plays no role but it is convenient to make this generalisation to support the cases
+`R = ℕ` and `R = ℤ` which both show up naturally. -/
+@[simps] def mk (p : add_submonoid M)
+  (hA : ∀ (a : A) {m : M}, m ∈ p → a • m ∈ p)
+  (hB : ∀ (b : B) {m : M}, m ∈ p → b • m ∈ p) : submodule (A ⊗[R] B) M :=
+{ carrier := p,
+  smul_mem' := λ ab m, tensor_product.induction_on ab
+   (λ hm, by simpa only [zero_smul] using p.zero_mem)
+   (λ a b hm, by simpa only [tensor_product.algebra.smul_def] using hA a (hB b hm))
+   (λ z w hz hw hm, by simpa only [add_smul] using p.add_mem (hz hm) (hw hm)),
+  .. p }
+
+lemma smul_mem (p : submodule (A ⊗[R] B) M) (a : A) {m : M} (hm : m ∈ p) : a • m ∈ p :=
+begin
+  suffices : a • m = a ⊗ₜ[R] (1 : B) • m, { exact this.symm ▸ p.smul_mem _ hm, },
+  simp [tensor_product.algebra.smul_def],
+end
+
+lemma smul_mem' (p : submodule (A ⊗[R] B) M) (b : B) {m : M} (hm : m ∈ p) : b • m ∈ p :=
+begin
+  suffices : b • m = (1 : A) ⊗ₜ[R] b • m, { exact this.symm ▸ p.smul_mem _ hm, },
+  simp [tensor_product.algebra.smul_def],
+end
+
+/-- Forgetting the `B` action, a `submodule` over `A ⊗[R] B` is just a `submodule` over `A`. -/
+@[simps] def to_submodule (p : submodule (A ⊗[R] B) M) : submodule A M :=
+{ carrier := p,
+  smul_mem' := smul_mem p,
+  .. p }
+
+/-- Forgetting the `A` action, a `submodule` over `A ⊗[R] B` is just a `submodule` over `B`. -/
+@[simps] def to_submodule' (p : submodule (A ⊗[R] B) M) : submodule B M :=
+{ carrier := p,
+  smul_mem' := smul_mem' p,
+  .. p }
+
+end algebra
+
+section ring
+
+variables (R S M : Type*) [ring R] [ring S]
+variables [add_comm_group M] [module R M] [module S M] [smul_comm_class R S M]
+
+/-- A `submodule` over `R ⊗[ℕ] S` is naturally also a `submodule` over the canonically-isomorphic
+ring `R ⊗[ℤ] S`. -/
+@[simps] def to_subbimodule_int (p : submodule (R ⊗[ℕ] S) M) : submodule (R ⊗[ℤ] S) M :=
+mk p.to_add_submonoid (smul_mem p) (smul_mem' p)
+
+/-- A `submodule` over `R ⊗[ℤ] S` is naturally also a `submodule` over the canonically-isomorphic
+ring `R ⊗[ℕ] S`. -/
+@[simps] def to_subbimodule_nat (p : submodule (R ⊗[ℤ] S) M) : submodule (R ⊗[ℕ] S) M :=
+mk p.to_add_submonoid (smul_mem p) (smul_mem' p)
```

Comment:
Do these generalize?

Reply:
I guess they do. I have added a commit expressing this.